datajunction-server/datajunction_server/construction/build_v2.py

```diff
@@ -179,202 +184,398 @@ async def get_measures_query(  # pylint: disable=too-many-locals
+                else Dialect.SPARK
+            )
+        )
+        return BuildCriteria(
+            dialect=dialect,
+            target_node_name=self.node_revision.name,
+        )
 
-    # Initialize the final query AST structure
-    final_ast = ast.Query(
-        select=ast.Select(
-            projection=[
-                ast.Column(
-                    ast.Name(col.alias_or_name.name),  # type: ignore
-                    _table=node_ast,
-                    _type=col.type,  # type: ignore
-                )
-                for col in node_ast.select.projection
-            ],
-            from_=ast.From(relations=[ast.Relation(node_alias)]),  # type: ignore
-        ),
-        ctes=[*node_ast.ctes, node_ast],
-    )
-    node_ast.ctes = []
-    node_ast = node_ast.to_cte(node_alias)
-    cte_mapping[node.name] = node_ast
-
-    # Start building the dimension joins and adding them to the CTEs
-    for _, dimension_join in dimension_node_joins.items():
-        join_path = dimension_join.join_path
-        requested_dimensions = list(dict.fromkeys(dimension_join.requested_dimensions))
-
-        for link in join_path:
-            link = cast(DimensionLink, link)
-            if all(
-                dim in link.foreign_keys_reversed for dim in requested_dimensions
-            ):  # pylint: disable=line-too-long # pragma: no cover
-                continue  # pragma: no cover
-
-            if link.dimension.name in cte_mapping:
-                dimension_join.node_query = cte_mapping[link.dimension.name]
-                continue
+    async def build(self) -> ast.Query:
+        """
+        Builds the node SQL with the requested set of dimensions, filter expressions,
+        order by, and limit clauses.
+
+        Build Strategy
+        ---------------
+        1. Recursively turn node references into query ASTs + apply any filters that can
+        be pushed down. If the node query has CTEs, unwind them into subqueries.
+        2. Initialize the final query with the node's query AST added to it as a CTE.
+        3. For any dimensions or filters requested for the node, determine if a join is
+        needed to bring in the dimension or filter. Keep track of all the necessary dimension
+        joins in a dict that maps dimension nodes to join paths.
+        4. For each of the necessary dimension joins, build the dimension node's query in the
+        same manner as above, recursively replacing any node references and pushing down requested
+        filters where possible.
+        5. Add each dimension node's query AST to the final query as a CTE.
+        6. Build the final query using the various CTEs. This does all the joins between the node
+        query AST and the dimension nodes' ASTs using the join logic from the dimension links.
+        7. Add all requested dimensions to the final select.
+        8. Add order by and limit to the final select (TODO)
+        """
+        physical_table = get_table_for_node(
```

Comment:
This would be cleaner as a property.
```py
@property
def physical_table(self):
    return get_table_for_node(
        self.node_revision,
        build_criteria=self._build_criteria,
    )
```